project/Component/NefMarkdown/NefMarkdown.swift

```diff
@@ -10,35 +10,77 @@ import Bow
 import BowEffects
 
 public struct Markdown {
-    private let render = Render()
+    public typealias MarkdownEnvironment = RenderMarkdownEnvironment<String>
+    typealias PageOutput  = RenderingOutput<String>.PageOutput
+    typealias PlaygroundOutput  = RenderingOutput<String>.PlaygroundOutput
+    typealias PlaygroundsOutput = RenderingOutput<String>.PlaygroundsOutput
     
     public init() {}
     
-    public func renderPage(content: String) -> EnvIO<RenderMarkdownEnvironment, RenderError, (rendered: String, ast: String)> {
-        render.renderPage(content: content).contramap(\RenderMarkdownEnvironment.renderEnvironment).map { rendered in
-            (rendered: rendered.output, ast: rendered.ast)
-        }^
+    public func page(content: String) -> EnvIO<MarkdownEnvironment, RenderError, (ast: String, rendered: String)> {
+        let env = EnvIO<MarkdownEnvironment, RenderError, MarkdownEnvironment>.var()
+        let rendered = EnvIO<MarkdownEnvironment, RenderError, PageOutput>.var()
+        
+        return binding(
+                 env <- ask(),
+            rendered <- env.get.render.page(content: content).contramap(\MarkdownEnvironment.renderEnvironment),
+        yield: (rendered: self.contentFrom(page: rendered.get), ast: rendered.get.ast))^
     }
-    
-    public func renderPage(content: String, filename: String, into output: URL) -> EnvIO<RenderMarkdownEnvironment, RenderError, (url: URL, ast: String, trace: String)> {
+
+    public func page(content: String, filename: String, into output: URL) -> EnvIO<MarkdownEnvironment, RenderError, (url: URL, ast: String, rendered: String)> {
         let file = output.appendingPathComponent(filename)
-        return render.renderPage(content: content, atFile: file).contramap(\RenderMarkdownEnvironment.renderEnvironment)
+        
+        let env = EnvIO<MarkdownEnvironment, RenderError, MarkdownEnvironment>.var()
+        let content = EnvIO<MarkdownEnvironment, RenderError, String>.var()
+        let rendered = EnvIO<MarkdownEnvironment, RenderError, PageOutput>.var()
+        
+        return binding(
+                 env <- ask(),
+             content <- env.get.fileSystem.readFile(atPath: file.path).mapLeft { _ in .renderPage(file) }.env(),
+            rendered <- env.get.render.page(content: content.get).contramap(\MarkdownEnvironment.renderEnvironment),
+                     |<-env.get.renderSystem.writePage(rendered.get, file).contramap(\MarkdownEnvironment.fileSystem).mapError { _ in .renderPage(file) },
```

Comment:
Each of the instructions in this comprehension is difficult to read on its own, as it needs adapting the dependencies and the error type, making it difficult to know what is going on. This can be avoided extracting the adaptation out to local inner functions and making the comprenhension cleaner:

```swift
binding(
  env <- ask(),
  content <- readFile(at: file.path, using: env.get.fileSystem),
  rendered <- render(content: content.get, using: env.get.render)
  |<- write(content: rendered.get, using: env.get.fileSystem,
  yield: (url: file, ast: rendered.get.ast, rendered: self.contentFrom(page: rendered.get)))^
```

This version lets us clearly see the process is read-render-write.

Reply:
Please, apply this to similar comprehensions.